test/Test-Stores.shared.js

```diff
@@ -1,192 +1,137 @@
-      const fileDeletedEvent = sinon.fake()
-      const req = { headers: { 'upload-length': 1000 }, url: this.storePath }
-
-      this.server.datastore.on(EVENTS.EVENT_FILE_DELETED, fileDeletedEvent)
-
-      const file = await this.server.datastore.create(req)
-      await this.server.datastore.remove({ file_id: file.id })
-      assert.equal(fileDeletedEvent.calledOnce, true)
-      return assert.rejects(this.server.datastore.remove({ file_id: file.id }))
+      await this.datastore.create(file)
+      return this.datastore.remove(file.id)
     })
   })
 }
 
 exports.shouldWriteUploads = function () {
   describe('write', function () {
-    it('should reject write streams that are not open yet', function () {
+    const file = new File('1234', `${this.testFileSize}`, undefined, 'filename d29ybGRfZG9taW5hdGlvbl9wbGFuLnBkZg==,is_confidential');
+
+    it('should reject write streams that can not be open', async function () {
       const stream = fs.createReadStream(this.testFilePath)
-      return this.server.datastore.write(stream, null, 0).should.be.rejectedWith(500)
+      return this.datastore.write(stream, 'doesnt_exist', 0).should.be.rejected()
     })
 
-    it('should open a stream, resolve the new offset, and emit upload complete', function (done) {
-      const uploadCompleteEvent = sinon.fake()
-      const req = {
-        headers: {
-          'upload-length': this.testFileSize.toString(),
-          'upload-metadata': 'foo bar',
-        },
-        url: this.storePath,
-      }
-
-      this.server.datastore.on(EVENTS.EVENT_UPLOAD_COMPLETE, uploadCompleteEvent)
-
+    it('should not reject whean readable stream has an error', async function () {
       const stream = fs.createReadStream(this.testFilePath)
-      const size = this.testFileSize
-      let id
-
-      stream.once('open', () => {
-        this.server.datastore
-          .create(req)
-          .then((file) => {
-            id = file.id
-            return this.server.datastore.write(stream, file.id, 0)
-          })
-          .then((offset) => {
-            assert.equal(offset, size)
-            assert.equal(uploadCompleteEvent.calledOnce, true)
-            return this.server.datastore.getOffset(id)
-          })
-          .then((stats) => {
-            assert.equal(stats.upload_length, size)
-          })
-          .then(done)
-          .catch(done)
-      })
+      return this.datastore.write(stream, 'doesnt_exist', 0).should.be.rejected()
     })
+
+    it('should write a stream and resolve the new offset', async function () {
+      await this.datastore.create(file);
+
+      const readable = fs.createReadStream(this.testFilePath);
+      const offset = await this.datastore.write(readable, file.id, 0);
+      assert.equal(offset, this.testFileSize);
+    });
+
+    // TODO: Does this test make sense? Should the store just report an error, and force the server to get offset via `getOffset`?
```

Comment:
It feels a bit off indeed when an underlying error occurred we still resolve an offset as if everything is fine. But not sure if `getOffset` still works after an error, depends on clean up I suppose

Reply:
As per specification server should try to save as mush as it can. If there is an error some data might be lost and probably an error reported. HEAD request should always succeed IMO and allow client to continue with upload. If it does not, then I'd say implementation is inadequate.

Reply:
I agree, it is probably best to not rely on `write` reporting the correct offset after an error.